File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        q = deque([root])
        layer = [] # list of int
        result = [] # list of list of int

        level_size = len(q)
        count = 0

        while q:
            node = q.popleft()
            layer.append(node.val)

            count += 1

            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

            if count == level_size:
                result.append(layer.copy())
                layer = []
                count = 0
                level_size = len(q)

                
        
        return result

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> l{};
        
        auto dfs = [&](this auto self, TreeNode* node) -> void{
            if (!node) return;

            self(node->left);
            l.push_back(node->val);
            self(node->right);
        };

        dfs(root);

        return l[k - 1];
    }
};
